Add Lock View documents (#5798)
pingcap/docs: information-schema/information-schema-deadlocks.md

The meaning of each column field in the `DEADLOCKS` table is as follows:

* `DEADLOCK_ID`: The ID of the deadlock event. When multiple deadlock errors exist in the table, you can use this column to distinguish rows that belong to different deadlock errors.
* `OCCUR_TIME`: The time when the deadlock error occurs.
* `RETRYABLE`: Whether the deadlock error can be retried. Currently, TiDB does not support collecting the information of the retryable deadlock error, so the value of this field is always `0`. For the description of retryable deadlock errors, see the [Retryable deadlock errors](#retryable-deadlock-errors) section.
* `TRY_LOCK_TRX_ID`: The ID of the transaction that tries to add lock, which is the `start_ts` of the transaction.
* `CURRENT_SQL_DIGEST`: The digest of the SQL statement currently being executed in the lock-adding transaction.
* `KEY`: The blocked key that the transaction tries to lock. This value of this field is displayed in the form of hexadecimal code.

> * `TRX_HOLDING_LOCK`: The ID of the transaction that currently holds the lock on the key and causes blocking, which is the `start_ts` of the transaction.

Author: Good question. What about "The ID of the transaction that currently holds the lock on the key and causes blocking. This ID is also the `start_ts` of the transaction."?

> * `TRX_HOLDING_LOCK`: The ID of the transaction that currently holds the lock on the key and causes blocking, which is the `start_ts` of the transaction.

Reviewer: Looks fine, I think.

> * `TRX_HOLDING_LOCK`: The transaction ID that currently holds the lock on the key and causes blocking, which is the transaction's `start_ts`.

Author: ```suggestion
* `TRX_HOLDING_LOCK`: The ID of the transaction that currently holds the lock on the key and causes blocking, which is the `start_ts` of the transaction.
```

> To adjust the number of deadlock events that can be accommodated in the `DEADLOCKS` table, adjust the [`pessimistic-txn.deadlock-history-capacity`](/tidb-configuration-file.md#deadlock-history-capacity) configuration in the TiDB configuration file. By default, the information of the recent 10 deadlock events is accommodated in the table..

Author: ```suggestion
To adjust the number of deadlock events that can be displayed in the `DEADLOCKS` table, adjust the [`pessimistic-txn.deadlock-history-capacity`](/tidb-configuration-file.md#deadlock-history-capacity) configuration in the TiDB configuration file. By default, the information of the recent 10 deadlock events is displayed in the table.
```

| DEADLOCK_ID | OCCUR_TIME                 | RETRYABLE | TRY_LOCK_TRX_ID    | CURRENT_SQL_DIGEST                                               | KEY                                    | TRX_HOLDING_LOCK   |
+-------------+----------------------------+-----------+--------------------+------------------------------------------------------------------+----------------------------------------+--------------------+
|           1 | 2021-06-04 08:22:38.765699 |         0 | 425405959304904707 | 22230766411edb40f27a68dadefc63c6c6970d5827f1e5e22fc97be2c4d8350d | 7480000000000000385F728000000000000002 | 425405959304904708 |
|           1 | 2021-06-04 08:22:38.765699 |         0 | 425405959304904708 | 22230766411edb40f27a68dadefc63c6c6970d5827f1e5e22fc97be2c4d8350d | 7480000000000000385F728000000000000001 | 425405959304904707 |
+-------------+----------------------------+-----------+--------------------+------------------------------------------------------------------+----------------------------------------+--------------------+
```

> Two rows of data are generated in the above table. The `DEADLOCK_ID` field of both rows is `1`, which means that the two rows of data contain the same deadlock error information. The first row shows that the transaction of the ID `425405959304904707` is blocked on the key of `"7480000000000000385F728000000000000002"` by the transaction of the ID `"425405959304904708"`. The second row shows that the transaction of the ID `"425405959304904708"` is blocked on the key of `"7480000000000000385F728000000000000001"` by the transaction of the ID `425405959304904707`, which constitutes mutual blocking and forms a deadlock.

Author: ```suggestion
Two rows of data are generated in the above table. The `DEADLOCK_ID` field of both rows is `1`, which means that the two rows of data display the same deadlock error information. The first row shows that the transaction of the ID `425405959304904707` is blocked on the key of `"7480000000000000385F728000000000000002"` by the transaction of the ID `"425405959304904708"`. The second row shows that the transaction of the ID `"425405959304904708"` is blocked on the key of `"7480000000000000385F728000000000000001"` by the transaction of the ID `425405959304904707`, which constitutes mutual blocking and forms a deadlock.
```

|           2 | 2021-06-04 08:22:56.795410 |         0 | 425405961664462854 | 22230766411edb40f27a68dadefc63c6c6970d5827f1e5e22fc97be2c4d8350d | 7480000000000000385F728000000000000003 | 425405961664462855 |
|           2 | 2021-06-04 08:22:56.795410 |         0 | 425405961664462855 | 22230766411edb40f27a68dadefc63c6c6970d5827f1e5e22fc97be2c4d8350d | 7480000000000000385F728000000000000001 | 425405961664462853 |
+-------------+----------------------------+-----------+--------------------+------------------------------------------------------------------+----------------------------------------+--------------------+
```

The `DEADLOCK_ID` column in the above query result shows that the first two rows together represent the information of a deadlock error, and the two transactions that wait for each other form a deadlock. The next three rows together represent another deadlock information, and the three transactions that wait in a cycle form a deadlock.

## Retryable deadlock errors

> **Note:**
>

> > Currently, TiDB does not support collecting retryable deadlock errors in `DEADLOCKS` table.

Author: ```suggestion
> Currently, TiDB does not support collecting retryable deadlock errors in the `DEADLOCKS` table.
```

> When transaction A is blocked by a lock already held by transaction B, and transaction B is directly or indirectly blocked by the lock held by the current transaction, a deadlock error will occur. In this deadlock, there might be two cases:

Author: @MyonKeminta PTAL

> When transaction A is blocked by a lock already held by transaction B, and transaction B is directly or indirectly blocked by the lock held by the current transaction, a deadlock error will occur. In this deadlock, there might be two cases:

Reviewer: yes

> When transaction A is blocked by a lock already held by transaction B, and transaction B is directly or indirectly blocked by the lock held by the current transaction, a deadlock error will occur. In this deadlock, there might be two cases:

Author: ```suggestion
When transaction A is blocked by a lock already held by transaction B, and transaction B is directly or indirectly blocked by the lock held by the current transaction A, a deadlock error will occur. In this deadlock, there might be two cases:
```

> + Case 1: Transaction B might (directly or indirectly) be blocked by a lock generated by a statement that has been executed before transaction A.

Reviewer: 事务 A 开始后到被阻塞前这段时间已经执行完成的语句

> + Case 1: Transaction B might (directly or indirectly) be blocked by a lock generated by a statement that has been executed before transaction A.

Author: @MyonKeminta @CharLotteiu PTAL
```suggestion
+ Case 1: Transaction B might be (directly or indirectly) blocked by a lock generated by a statement that has been executed after transaction A starts and before transaction A gets blocked.
```

> + Case 1: Transaction B might (directly or indirectly) be blocked by a lock generated by a statement that has been executed before transaction A.

Reviewer 2: LGTM

> + Case 2: Transaction B might also be blocked by the statement currently being executed in Transaction A.

Author: ```suggestion
+ Case 2: Transaction B might also be blocked by the statement currently being executed in transaction A.
```

In case 1, TiDB will report a deadlock error to the client of transaction A and terminate the transaction.

In case 2, the statement currently being executed in transaction A will be automatically retried in TiDB. For example, suppose that transaction A executes the following statement:

{{< copyable "sql" >}}

```sql
update t set v = v + 1 where id = 1 or id = 2;
```

Transaction B executes the following two statements successively.

{{< copyable "sql" >}}

```sql
update t set v = 4 where id = 2;
update t set v = 2 where id = 1;
```

Then if transaction A locks the two rows with `id = 1` and `id = 2`, and the two transactions run in the following sequence:

1. Transaction A locks the row with `id = 1`.
2. Transaction B executes the first statement and locks the row with `id = 2`.
3. Transaction B executes the second statement and tries to lock the row with `id = 1`, which is blocked by transaction A.

> 4. Transaction A tries to lock the row with `id = 2` and is blocked by B, which forms a deadlock.

Author: ```suggestion
4. Transaction A tries to lock the row with `id = 2` and is blocked by transaction B, which forms a deadlock.
```

+--------------------+---------------------+------+------+---------+-------+
| Field              | Type                | Null | Key  | Default | Extra |
+--------------------+---------------------+------+------+---------+-------+
| INSTANCE           | varchar(64)         | YES  |      | NULL    |       |
| DEADLOCK_ID        | bigint(21)          | NO   |      | NULL    |       |
| OCCUR_TIME         | timestamp(6)        | YES  |      | NULL    |       |
| RETRYABLE          | tinyint(1)          | NO   |      | NULL    |       |
| TRY_LOCK_TRX_ID    | bigint(21) unsigned | NO   |      | NULL    |       |
| CURRENT_SQL_DIGEST | varchar(64)         | YES  |      | NULL    |       |
| KEY                | text                | YES  |      | NULL    |       |
| TRX_HOLDING_LOCK   | bigint(21) unsigned | NO   |      | NULL    |       |
+--------------------+---------------------+------+------+---------+-------+
```

## SQL Digest

The `DEADLOCKS` table records the SQL digest but not the original SQL statement.

SQL digest is the hash value after the SQL normalization. To find the original SQL statement corresponding to the SQL digest, perform one of the following operations:

> - For the statements executed on the current TiDB node in the recent period of time, you can find the corresponding orginal SQL statement from the SQL digest in `STATEMENTS_SUMMARY` or `STATEMENTS_SUMMARY_HISTORY`.
> - For the statements executed on all TiDB nodes in the entire cluster in the recent period of time, you can find the corresponding SQL statement from the SQL Digest in `CLUSTER_STATEMENTS_SUMMARY` or `CLUSTER_STATEMENTS_SUMMARY_HISTORY`.

Author: ```suggestion
- For the statements executed on the current TiDB node in the recent period of time, you can find the corresponding original SQL statement in the `STATEMENTS_SUMMARY` or `STATEMENTS_SUMMARY_HISTORY` table according to the SQL digest.
- For the statements executed on all TiDB nodes in the entire cluster in the recent period of time, you can find the corresponding SQL statement in the `CLUSTER_STATEMENTS_SUMMARY` or `CLUSTER_STATEMENTS_SUMMARY_HISTORY` table according to the SQL digest.
```
